torchao/dtypes/fpx/fpx.py

```diff
@@ -347,101 +354,90 @@ def from_scaled_tc_fpx(tensor: Tensor, ebits: int, mbits: int, scale=None) -> Te
+        self.packed_fpx_data = packed_fpx_data
         self.ebits = ebits
         self.mbits = mbits
 
     def __tensor_flatten__(self):
-        return ["fpx_data", "scale"], [self.ebits, self.mbits]
+        return ["packed_fpx_data"], [self.ebits, self.mbits]
 
     @classmethod
-    def __tensor_unflatten__(cls, tensor_data_dict, tensor_attributes, outer_size=None, outer_stride=None):
-        return cls(tensor_data_dict["fpx_data"], tensor_data_dict["scale"], *tensor_attributes)
-
-    @classmethod
-    def from_float(cls, input_float: Tensor, ebits: int, mbits: int):
-        fpx_data, scale = to_scaled_tc_fpx(input_float, ebits, mbits)
-        return cls(fpx_data, scale, ebits, mbits)
-
-    def dequantize(self, output_dtype=None):
-        output_dtype = output_dtype or torch.get_default_dtype()
-        return from_scaled_tc_fpx(self.fpx_data, self.ebits, self.mbits, self.scale).to(output_dtype)
+    def __tensor_unflatten__(cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride):
+        return cls(tensor_data_dict["packed_fpx_data"], *tensor_attributes)
 
     def __repr__(self):
         dtype = f"fp{1 + self.ebits + self.mbits}_e{self.ebits}m{self.mbits}"
         return (
             f"{self.__class__.__name__}(dtype={dtype}, shape={self.shape}, "
+            f"data={self.get_plain()}, "
             f"device={self.device}, requires_grad={self.requires_grad})"
         )
 
     def _apply_fn_to_data(self, fn):
         return self.__class__(
-            fn(self.fpx_data),
-            fn(self.scale),
+            fn(self.packed_fpx_data),
             self.ebits,
             self.mbits,
         )
 
-@QuantLlmLinearWeight.implements(torch.nn.functional.linear)
-def _(func, types, args, kwargs):
-    act = args[0]
-    weight = args[1]
-    bias = args[2] if len(args) >= 3 else None
-    assert isinstance(weight, QuantLlmLinearWeight)
-
-    out_dim, in_dim = weight.shape
-    act_reshaped = act.view(-1, in_dim).half()
-
-    # https://github.com/microsoft/DeepSpeed/blob/3a3a6db3332e339cc9fd94efd4982f6d60635a3d/deepspeed/inference/v2/kernels/core_ops/cuda_linear/cuda_linear.py
-    bsize = act_reshaped.shape[0]
-    splitK = _SPLIT_K_MAP[(bsize - 1) // 64].get(out_dim, 1) if bsize <= 768 else 1
-
-    out = quant_llm_linear(
-        weight.ebits,
-        weight.mbits,
-        act_reshaped,
-        weight.fpx_data,
-        weight.scale,
-        splitK=splitK,
-    )
-
-    if bias is not None:
-        out += bias
+    def get_plain(self):
+        """Returns the unpacked uint8 fpx data
+        """
+        return unpack_tc_fpx(self.packed_fpx_data, 1 + self.ebits + self.mbits)
 
-    return out.view(*act.shape[:-1], out_dim).to(act.dtype)
+    @classmethod
+    def from_uint8(cls, fpx_data: Tensor, ebits: int, mbits: int):
+        packed_fpx_data = pack_tc_fpx(fpx_data, 1 + ebits + mbits)
+        return cls(packed_fpx_data, ebits, mbits)
```

Comment:
My concern with this is that `pack_tc_fpx()` and `unpack_tc_fpx()` actually packs data in a layout optimized for Tensor Cores, as outlined in FP6-LLM paper (the idea should be similar to tinygemm TensorCoreLayout, though I don't think they are exactly identical). So I think this should be under a Layout class instead. This wouldn't be a generic FPx data (which would raise the question, what should a generic, placeholder FPx data layout should look like).

Reply:
I see.. OK in that case I can rename this, what is a good name? `FpxQuantLLMLayout`?

Reply:
That should be fine! `FpxLLMLayout` is also possible. No big preference on this.
Then we can mention the original paper+repo in the docstring.
Also, I think it's a good idea to add a short docstring about the layout of returned value from `.get_plain()` and expected layout of `.from_uint8(fpx_data)` i.e. unpacked FPx data with padding - `00SEEEMM` for FP6 E3M2. I think MX dtypes have some good docstring for this.

Reply:
makes sense, updated

Reply:
Yea the problem is we don't know what FPx plain layout should look like for odd-bitwidth. For FP4, it's natural to pack 2 values in a byte. FP1 and FP2 does not make sense so I don't think they will exist 😄. For the rest, FP3, FP5, FP6, FP7, it's tricky to define their plain layout.

Reply:
I thought grouping things with 1, 2, 4 bit shards feels like the standard approach? this is how uintx is packed as well?

Reply:
I wouldn't say it's "standard" (I might be wrong). @vayuda adopts that idea from FP6-LLM also 😅.

Reply:
ah I see, OK, maybe I should change uintx to say it's optimized for something as well, instead of saying it's native..